tests/server.py

```diff
@@ -588,48 +589,48 @@ class FakeTransport:
     If you want bidirectional communication, you'll need two instances.
     """
 
-    other = attr.ib()
+    other: IProtocol
     """The Protocol object which will receive any data written to this transport.
 
     :type: twisted.internet.interfaces.IProtocol
     """
 
-    _reactor = attr.ib()
+    _reactor: IReactorTime
     """Test reactor
 
     :type: twisted.internet.interfaces.IReactorTime
     """
 
-    _protocol = attr.ib(default=None)
+    _protocol: Optional[IProtocol] = None
     """The Protocol which is producing data for this transport. Optional, but if set
     will get called back for connectionLost() notifications etc.
     """
 
-    _peer_address: Optional[IAddress] = attr.ib(default=None)
+    _peer_address: Optional[IAddress] = None
     """The value to be returned by getPeer"""
 
-    _host_address: Optional[IAddress] = attr.ib(default=None)
+    _host_address: Optional[IAddress] = None
     """The value to be returned by getHost"""
 
     disconnecting = False
     disconnected = False
     connected = True
-    buffer = attr.ib(default=b"")
-    producer = attr.ib(default=None)
-    autoflush = attr.ib(default=True)
+    buffer: bytes = attr.Factory(bytes)
+    producer: Optional[IPushProducer] = None
+    autoflush: bool = True
 
     def getPeer(self) -> Optional[IAddress]:
         return self._peer_address
 
     def getHost(self) -> Optional[IAddress]:
         return self._host_address
 
-    def loseConnection(self, reason=None):
+    def loseConnection(self, reason: Optional[Failure] = None) -> None:
         if not self.disconnecting:
             logger.info("FakeTransport: loseConnection(%s)", reason)
             self.disconnecting = True
             if self._protocol:
-                self._protocol.connectionLost(reason)
+                self._protocol.connectionLost(reason)  # type: ignore[arg-type]
```

Comment:
Two things I noticed here:

- According to https://docs.twistedmatrix.com/en/stable/api/twisted.internet.interfaces.ITransport.html#loseConnection , `loseConnection` is expected to be callable with zero arguments (which it is, since `reason` has a default). Should we remove the `reason` argument alltogether?
- According to https://github.com/twisted/twisted/blob/7f421f9c55e41c36a4e1032fe85b5f0f35fb5bc1/src/twisted/internet/protocol.py#L571 it looks like we're not supposed to pass `None` to `connectionLost`. Should we make up a dummy `Failure` object to be good citizens?

Reply:
Oh I think someone got confused at some point... `connectionLost(reason)` takes a failure, but not `loseConnection`. (Yes, confusing names...)